### ISSUE TITLE
Best practices warns on vkWaitForFences with VK_TIMEOUT=0?

### ISSUE THREAD

**Issue by haasn** [KhronosGroup/Vulkan-ValidationLayers#1788]
I'm not entirely sure what the purpose of this check is, but as it stands, it's reporting warnings such as

```
vk [Validation] 0: Validation Warning: [ UNASSIGNED-BestPractices-NonSuccess-Result ] Object 0: handle = 0x555555632820, type = VK_OBJECT_TYPE_INSTANCE; | MessageID = 0x8928392f | vkWaitForFences(): Returned error VK_TIMEOUT. (obj 0x555555632820 (VkInstance), loc 0x8928392f)
```

For a call where `timeout` was specified as `0`, which by the spec does not block but merely returns the current status of the fence. I feel like the intent here may have been to only warn for actual timeouts?

Is there a legitimate reason to avoid this sort of "non-blocking polling" of VkFences, or is this a false positive?

For context, the usage that triggers the warning is a code path in which the fence is opportunistically checked with `timeout=0` to see if the command has already succeeded "in the meantime", and if this is not the case, the rendering pipeline is flushed and the poll is retried with `timeout=UINT64_MAX` to block until it succeeds. I feel like this usage is justified, even though I could theoretically avoid it by using better logic to determine whether I need to flush or not.

**Comment by mark-lunarg** [comment]
@haasn, you were right!  Pushed the fix to the repo, but leaving this open until tests can be added.  Thanks for the quick report and sorry for the trouble.

**Comment by haasn** [comment]
Current master fixes the confusion between "error" and "non-success", but I still get the same warning (albeit with the wording reaplced). In this logic:

```cpp
    auto success = std::find(success_codes.begin(), success_codes.end(), result);
    if (success != success_codes.end()) {
        LogWarning(instance, kVUID_BestPractices_Error_Result, "%s(): Returned non-success return code %s.", api_name,
                   string_VkResult(result));
    }
```

The check was probably intended to be `if (success == success_codes.end())`, i.e. only printing the warning if `result` was *not* contained in `success_codes`.

**Comment by mark-lunarg** [comment]
Ah.  See, should've written the tests first.  Anyway, doing that now and will get in a PR soon.  Thanks!

**Comment by mark-lunarg** [comment]
@haasn, the check was correct but the messages were backwards!  Just to be clear, the intent IS to print a message to let you know that a successful return code was received, but that it was NOT `VK_SUCCESS`.  This is because special action may need to be taken and folks can miss it.  I have a PR going up shortly, but wondered if it was the prospect of getting ANY kind of a warning for a 'success' code that was objectionable to you...

**Comment by haasn** [comment]
Ah, that justification makes sense. Although it does raise the question of how to squelch this warning for the cases where you *do* successfully handle the success return code. It seems like the way this logic is designed currently will end up necessarily spamming users with false positives.

Maybe the verbosity level of this can be reduced from warning down to something more informational?

**Comment by mark-lunarg** [comment]
Exactly!  Am considering making the non-success success codes an `INFORMATION `type instead of a `WARNING`, but that doesn't really help if you have both enabled. If you have a callback set up, you can just toss any message with that `messageIdNumber` and pretend you never saw it.  We'll have some tools in the near term to make this easier, but those are our choices ATM.  What do you think?

**Comment by haasn** [comment]
I think making it `INFORMATION` rather than `WARNING` is a good idea. I do have both enabled, but warnings and errors are made much more visible by my debug callback.

**Comment by mark-lunarg** [comment]
OK, I'll do that for now, but I can't promise that they'll get changed to warnings in the future!  DebugUtils has `MessageSeverity `and `MessageType`, but these don't help at all here, where we really need something like 'just letting you know...', 'a helpful hint', 'you might want to take a look',  'Did you REALLY mean to do that', and 'I don't think this is a good idea at all'.  Maybe we can inject keywords into the message strings, or VUIDs, to help with this.
